### ISSUE TITLE
hfst-lexc segfaults when using foma backend

### ISSUE THREAD

**Issue by snomos** [hfst/hfst#316]
Lexc code to repeat:

```
LEXICON Root
ARABIC ;

LEXICON ARABIC
< [1|2|3|4|5|6|7|8|9|%0] > # ;
```

Expected result: a simple transducer accepting the numbers 0-9. Actual result:

```
$ /usr/local/bin/hfst-lexc --format=foma -o test.hfst test.lexc 
Root...1 ARABIC...Segmentation fault: 11
$
```

This is not the case when using another backend:

```
$ /usr/local/bin/hfst-lexc -o test.hfst test.lexc 
/usr/local/bin/hfst-lexc: warning: Defaulting to OpenFst tropical type
Root...1 ARABIC...
$ hfst-lookup test.hfst 
hfst-lookup: warning: It is not possible to perform fast lookups with OpenFST, std arc, tropical semiring format automata.
Using HFST basic transducer format and performing slow lookups
> 1
1	1	0,000000
```

It is also not an issue when compiling the lexc code directly in Foma:

```
$ foma
Foma, version 0.9.18alpha
Copyright © 2008-2014 Mans Hulden
This is free software; see the source code for copying conditions.
There is ABSOLUTELY NO WARRANTY; for details, type "help license"

Type "help" to list all commands available.
Type "help <topic>" or help "<operator>" for further help.

foma[0]: read lexc test.lexc
Root...1, ARABIC...1
Building lexicon...
Determinizing...
Minimizing...
Done!
623 bytes. 2 states, 10 arcs, 10 paths.
foma[1]: quit
```

Tested against commit e6c6e82d487927f8dcc143f8d63fbd22b4a26110. Tested on OSX 10.11.4.

**Comment by TinoDidriksen** [comment]
This is not isolated to OS X. On Ubuntu, valgrind shows many many errors, they just don't result in a segfault.

```
$ valgrind -v --leak-check=full hfst-lexc --format=foma -o test.hfst test.lexc
ERROR SUMMARY: 92 errors from 84 contexts (suppressed: 0 from 0)
```

```
$ valgrind -v --leak-check=full hfst-lexc -o test.hfst test.lexc
ERROR SUMMARY: 4 errors from 4 contexts (suppressed: 0 from 0)
```

**Comment by eaxelson** [comment]
Changed 'delete' to 'free' when deallocating memory allocated with 'malloc' in revision d15b4be. This should fix the issue.

**Comment by snomos** [comment]
Sorry to inform you that it still segfaults:

$ ll tools/src/hfst-lexc
-rwxr-xr-x+ 1 smo036  1907360568   6,2K 26 aug 13:49 tools/src/hfst-lexc*
$ tools/src/hfst-lexc --format=foma -o test.hfst ~/langtech/test.lexc 
Root...1 ARABIC...Segmentation fault: 11
$ git pull
Already up-to-date.

**Comment by eaxelson** [comment]
The problem comes from using strdup with standard C99. Fixed in revision 6dc1d8f.
